Fix UniprotIO error on mass spec

laserson opened the pull request:
Some mass spec location tags (`begin` and `end`) don't have `position` attributes.  For example, one of them has `status="unknown".  This causes a KeyError.`

peterjc commented:
Can you list some UniProt IDs where this happens? I'd like to see the data (plain text and XML) and pick one for a unit test (ideally a small file).

laserson commented:
I think it was the 28th record into the SwissProt XML file: P84001

peterjc commented:
That's a single accession, and the feature location looks OK by eye:

* http://www.uniprot.org/uniprot/P84001
* http://www.uniprot.org/uniprot/P84001.txt
* http://www.uniprot.org/uniprot/P84001.xml

Do you have a URL you could share?

laserson commented:
This line in the XML file broke for me on `master`:

```
<end status="unknown"/>
```

laserson commented:
Running

```bash
wget http://www.uniprot.org/uniprot/P84001.xml
python -c 'from Bio.SeqIO import parse; parse("P84001.xml", "uniprot-xml").__next__()'
```

gives me the output

```
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/Users/laserson/miniconda3/lib/python3.5/site-packages/Bio/SeqIO/__init__.py", line 600, in parse
    for r in i:
  File "/Users/laserson/miniconda3/lib/python3.5/site-packages/Bio/SeqIO/UniprotIO.py", line 79, in UniprotIterator
    yield Parser(elem, alphabet=alphabet, return_raw_comments=return_raw_comments).parse()
  File "/Users/laserson/miniconda3/lib/python3.5/site-packages/Bio/SeqIO/UniprotIO.py", line 504, in parse
    _parse_comment(element)
  File "/Users/laserson/miniconda3/lib/python3.5/site-packages/Bio/SeqIO/UniprotIO.py", line 278, in _parse_comment
    end = int(list(loc_element.getiterator(NS + 'end'))[0].attrib['position'])
KeyError: 'position'
```

peterjc commented:
Confirmed - same for me running Biopython from git on Mac OS.

laserson commented:
I should say I am running Python 3.5 and Biopython abc9bed3b5cc4180b88565a8c6dbfe13e81ec051

laserson commented:
I can try to add a test myself but I'm not familiar with your infra.

peterjc commented:
Expected behaviour would generally match parsing the plain text file,

```
$ curl -L -O http://www.uniprot.org/uniprot/P84001.txt
$ python -c "from Bio import SeqIO; print(SeqIO.read('P84001.txt', 'swiss'))"
```

It looks like this is not the feature location that is breaking (which I was sure I'd tested), but a structured comment (which we likely don't handle fully in the plain text).

peterjc commented:
See the testing chapter in our tutorial, basically using ``unittest`` via ``Tests/run_tests.py``

peterjc commented:
The test looks good, but could you add another ``self.assertEqual(...)`` checking this problematic comment location?

peterjc commented:
Sorry, two more queries:

Are you happy to be thanked by name in the ``NEWS`` (alphabetical by first name for each release, with `` (first contribution)`` appended where appropriate) and ``CONTRIB`` files (alphabetically by surname, using GitHub profiles if known)? You can add yourself as another commit if you want, or I'd do as as part of merging the pull request.

Are you happy to dual-license your contributions to Biopython under BOTH the current Biopython License Agreement and the 3-clause BSD licence as per #898? Thanks!

laserson commented:
Rebased the code changes and added a commit updating my info in CONTRIB.  Feel free to add me to NEWS as well.  I'm fine with the license.  (Does this comment function as a CLA? 😛 )

peterjc commented:
We're currently collecting licensing agreements via GitHub and/or our public email list. If you have thoughts on CLA best practice, input is welcome, e.g. comments on #904

laserson commented:
Done.  My previous company had an actual contract that needed to be signed.  I'd check to see how Apache handles this.

laserson commented:
Anything else to do before merge?

peterjc commented:
I've not had time to see how this compares to the "swiss" format parser - is the same information there and parsed into the same data structure?

laserson commented:
The mass spec comment is not parsed using the `swiss` parser.

In the XML version, the `annotations` field contains
```
{'accessions': ['P84001'],
 'alternativeName_fullName': ['Venom protein ANC29C0'],
 'comment_function': ['Possible neurotoxin.'],
 'comment_massspectrometry': ['undefined:9571|Electrospray'],
 'comment_subcellularlocation_location': ['Secreted'],
 'comment_tissuespecificity': ['Expressed by the venom gland.'],
 'created': '2004-07-19',
 'dataset': 'Swiss-Prot',
 'key': ['1', '2', '3'],
 'keywords': ['Direct protein sequencing', 'Neurotoxin', 'Secreted', 'Toxin'],
 'modified': '2015-01-07',
 'organism': 'Ancylometes sp. (South American fishing spider)',
 'proteinExistence': ['evidence at protein level'],
 'recommendedName_fullName': ['U3-ctenitoxin-Asp1a'],
 'recommendedName_shortName': ['U3-CNTX-Asp1a'],
 'references': [Reference(title='Protein ANC29C0 from venom of South American fishing spider (Ancylometes spp.) has sequence similarities to neurotoxic peptide Caeron precursor from Bark spider (Caerostris extrusa).', ...)],
 'sequence_checksum': '3B06738CA0940B53',
 'sequence_fragment': 'single',
 'sequence_length': 50,
 'sequence_mass': 5679,
 'sequence_modified': '2004-07-19',
 'sequence_version': 1,
 'taxonomy': ['Eukaryota',
  'Metazoa',
  'Ecdysozoa',
  'Arthropoda',
  'Chelicerata',
  'Arachnida',
  'Araneae',
  'Araneomorphae',
  'Entelegynae',
  'Lycosoidea',
  'Ctenidae',
  'Ancylometes'],
 'type': ['ECO:0000269', 'ECO:0000303', 'ECO:0000305'],
 'version': 25}
```
while using the `swiss` parser, the `annotation` field gets
```
{'accessions': ['P84001'],
 'comment': 'FUNCTION: Possible neurotoxin. {ECO:0000303|Ref.1}.\nSUBCELLULAR LOCATION: Secreted {ECO:0000269|Ref.1}.\nTISSUE SPECIFICITY: Expressed by the venom gland. {ECO:0000269|Ref.1}.\nMASS SPECTROMETRY: Mass=9571; Method=Electrospray; Range=1-?; Evidence={ECO:0000269|Ref.1};',
 'date': '19-JUL-2004',
 'date_last_annotation_update': '07-JAN-2015',
 'date_last_sequence_update': '19-JUL-2004',
 'keywords': ['Direct protein sequencing', 'Neurotoxin', 'Secreted', 'Toxin'],
 'ncbi_taxid': ['280265'],
 'organism': 'Ancylometes sp. (South American fishing spider)',
 'references': [Reference(title='Protein ANC29C0 from venom of South American fishing spider (Ancylometes spp.) has sequence similarities to neurotoxic peptide Caeron precursor from Bark spider (Caerostris extrusa).', ...)],
 'taxonomy': ['Eukaryota',
  'Metazoa',
  'Ecdysozoa',
  'Arthropoda',
  'Chelicerata',
  'Arachnida',
  'Araneae',
  'Araneomorphae',
  'Entelegynae',
  'Lycosoidea',
  'Ctenidae',
  'Ancylometes']}
```

peterjc commented:
OK, good - no mismatch, but a future enhancement count add this to the plain text parser.

peterjc commented:
Rebased to the master - thank you!
